# **Classification**

Loading the feature-engineered dataset

In [4]:
import pandas as pd

df = pd.read_csv("feature_engineered_dataset.csv")

print("Dataset shape:", df.shape)
df.head()

Dataset shape: (388233, 30)


,age,gender,marital_status,education,monthly_salary,employment_type,years_of_employment,company_type,house_type,monthly_rent,...,bank_balance,emergency_fund,emi_scenario,requested_amount,requested_tenure,max_monthly_emi,total_monthly_expenses,disposable_income,expense_to_income_ratio,emi_eligibility
0,38.0,2,0,3,82600.0,1,0.9,2,2,20000.0,...,3998,70200.0,3,850000.0,15,500.0,27200.0,55400.0,0.329298,2
1,38.0,2,0,0,21500.0,1,7.0,1,0,0.0,...,11524,26900.0,0,128000.0,19,700.0,6500.0,15000.0,0.302326,2
2,38.0,5,0,3,86100.0,1,5.8,4,1,0.0,...,8679,324200.0,1,306000.0,16,27775.0,10200.0,75900.0,0.118467,0
3,58.0,2,0,1,66800.0,1,2.2,2,1,0.0,...,5844,178100.0,4,304000.0,83,16170.0,17600.0,49200.0,0.263473,0
4,48.0,2,0,3,57300.0,1,3.4,2,0,0.0,...,11945,28200.0,2,252000.0,7,500.0,34300.0,23000.0,0.598604,2


Check the target

In [5]:
print(df["emi_eligibility"].value_counts())
print(df["emi_eligibility"].value_counts(normalize=True))

emi_eligibility
2    300086
0     71380
1     16767
Name: count, dtype: int64
emi_eligibility
2    0.772953
0    0.183859
1    0.043188
Name: proportion, dtype: float64


Separate features and target

In [6]:
X = df.drop("emi_eligibility", axis=1)
y = df["emi_eligibility"]

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (388233, 29)
y shape: (388233,)


In [7]:
print("df shape:", df.shape)
print("X shape:", X.shape)
print("y shape:", y.shape)

df shape: (388233, 30)
X shape: (388233, 29)
y shape: (388233,)


In [8]:
# Recreate X and y from the CURRENT df

X = df.drop("emi_eligibility", axis=1)
y = df["emi_eligibility"]

print("df shape:", df.shape)
print("X shape:", X.shape)
print("y shape:", y.shape)

print("\nTarget distribution:")
print(y.value_counts(dropna=False))

print("\nTarget percentage:")
print(y.value_counts(normalize=True, dropna=False) * 100)

df shape: (388233, 30)
X shape: (388233, 29)
y shape: (388233,)

Target distribution:
emi_eligibility
2    300086
0     71380
1     16767
Name: count, dtype: int64

Target percentage:
emi_eligibility
2    77.295336
0    18.385866
1     4.318798
Name: proportion, dtype: float64


Split the data

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (310586, 29)
X_test: (77647, 29)
y_train: (310586,)
y_test: (77647,)


In [10]:
print("Total rows in df:", len(df))
print("Training rows:", len(X_train))
print("Testing rows:", len(X_test))

Total rows in df: 388233
Training rows: 310586
Testing rows: 77647


In [11]:
print("Missing target values:", y.isna().sum())
print("Total rows:", len(y))

print("\nTarget distribution:")
print(y.value_counts(dropna=False))

Missing target values: 0
Total rows: 388233

Target distribution:
emi_eligibility
2    300086
0     71380
1     16767
Name: count, dtype: int64


In [12]:
print("Rows with missing target:")
print(df[df["emi_eligibility"].isna()].head())

Rows with missing target:
Empty DataFrame
Columns: [age, gender, marital_status, education, monthly_salary, employment_type, years_of_employment, company_type, house_type, monthly_rent, family_size, dependents, school_fees, college_fees, travel_expenses, groceries_utilities, other_monthly_expenses, existing_loans, current_emi_amount, credit_score, bank_balance, emergency_fund, emi_scenario, requested_amount, requested_tenure, max_monthly_emi, total_monthly_expenses, disposable_income, expense_to_income_ratio, emi_eligibility]
Index: []

[0 rows x 30 columns]


# Import Logistic Regression

In [13]:
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(
    max_iter=1000,
    random_state=42
)

print("Logistic Regression model created!")

Logistic Regression model created!


In [14]:
# Train it
logistic_model.fit(X_train, y_train)

print("Logistic Regression training completed!")

Logistic Regression training completed!


c:\Users\avani\Documents\EMIPredict-AI\emi_env\Lib\site-packages\sklearn\linear_model\_logistic.py:599: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [15]:
# Make predictions
y_pred_logistic = logistic_model.predict(X_test)

print("Predictions completed!")
print("Number of predictions:", len(y_pred_logistic))

Predictions completed!
Number of predictions: 77647


In [16]:
# Calculate accuracy
from sklearn.metrics import accuracy_score

logistic_accuracy = accuracy_score(y_test, y_pred_logistic)

print("Logistic Regression Accuracy:", logistic_accuracy)
print("Accuracy (%):", logistic_accuracy * 100)

Logistic Regression Accuracy: 0.8828802143031926
Accuracy (%): 88.28802143031926


In [17]:
# Full evaluation of Logistic Regression
from sklearn.metrics import classification_report, confusion_matrix

print("Classification Report:")
print(classification_report(y_test, y_pred_logistic))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_logistic))

Classification Report:
              precision    recall  f1-score   support

           0       0.77      0.75      0.76     14276
           1       0.18      0.01      0.01      3353
           2       0.91      0.96      0.94     60018

    accuracy                           0.88     77647
   macro avg       0.62      0.57      0.57     77647
weighted avg       0.85      0.88      0.86     77647


Confusion Matrix:
[[10667    57  3552]
 [ 1010    22  2321]
 [ 2112    42 57864]]


**overall accuracy is 88.29%, but the model is poor at detecting class 1 That means Logistic Regression is detecting almost none of class 1. So 88.29% accuracy is misleading because class 2 is the dominant class.**

# Random Forest

In [18]:
# Train Random Forest
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

print("Random Forest model created!")

Random Forest model created!


In [19]:
# train
rf_model.fit(X_train, y_train)

print("Random Forest training completed!")

Random Forest training completed!


In [20]:
# predict
y_pred_rf = rf_model.predict(X_test)

print("Random Forest predictions completed!")

Random Forest predictions completed!


In [21]:
# Calculate accuracy
from sklearn.metrics import accuracy_score

rf_accuracy = accuracy_score(y_test, y_pred_rf)

print("Random Forest Accuracy:", rf_accuracy)
print("Accuracy (%):", rf_accuracy * 100)

Random Forest Accuracy: 0.9523870851417312
Accuracy (%): 95.23870851417313


In [22]:
# Full evaluation
print("Random Forest Classification Report:")
print(classification_report(y_test, y_pred_rf))

print("\nRandom Forest Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))

Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.97      0.94     14276
           1       0.83      0.07      0.13      3353
           2       0.96      1.00      0.98     60018

    accuracy                           0.95     77647
   macro avg       0.90      0.68      0.68     77647
weighted avg       0.95      0.95      0.94     77647


Random Forest Confusion Matrix:
[[13806    38   432]
 [ 1148   228  1977]
 [   93     9 59916]]


**Random Forest is much better overall than Logistic Regression.However, there is still a major issue: Class 1 (High_Risk) recall is only 0.07.**

**That means Random Forest correctly identifies only about 7% of the actual High_Risk cases.
So we should not stop at 95.24% accuracy.**

**Let's train a class-weighted Random Forest so mistakes on minority classes receive more importance.**

# Balanced Random Forest

In [23]:
# Train weighted Random Forest
rf_balanced = RandomForestClassifier(
    n_estimators=100,
    class_weight="balanced",
    random_state=42,
    n_jobs=-1
)

print("Balanced Random Forest created!")

Balanced Random Forest created!


In [24]:
# Train
rf_balanced.fit(X_train, y_train)

print("Balanced Random Forest training completed!")

Balanced Random Forest training completed!


In [25]:
# Predict
y_pred_rf_balanced = rf_balanced.predict(X_test)

print("Predictions completed!")

Predictions completed!


In [26]:
# Evaluate
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

rf_balanced_accuracy = accuracy_score(y_test, y_pred_rf_balanced)

print("Balanced Random Forest Accuracy:",
      rf_balanced_accuracy)

print("Accuracy (%):",
      rf_balanced_accuracy * 100)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_rf_balanced))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf_balanced))

Balanced Random Forest Accuracy: 0.9523870851417312
Accuracy (%): 95.23870851417313

Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.98      0.92     14276
           1       0.61      0.26      0.36      3353
           2       0.98      0.98      0.98     60018

    accuracy                           0.95     77647
   macro avg       0.82      0.74      0.76     77647
weighted avg       0.95      0.95      0.95     77647


Confusion Matrix:
[[14014   132   130]
 [ 1703   866   784]
 [  534   414 59070]]


**The accuracy stayed at 95.24%, but the important improvement is the minority class.**

# Gradient Boosting

In [27]:
# Create the model
from sklearn.ensemble import HistGradientBoostingClassifier

hgb_model = HistGradientBoostingClassifier(
    max_iter=100,
    learning_rate=0.1,
    max_leaf_nodes=31,
    random_state=42
)

print("HistGradientBoosting model created!")

HistGradientBoosting model created!


In [28]:
# Train
hgb_model.fit(X_train, y_train)

print("HistGradientBoosting training completed!")

HistGradientBoosting training completed!


In [29]:
print("Number of features:", len(X.columns))

print("\nColumns:")
print(X.columns.tolist())

print("\nData types:")
print(X.dtypes)

Number of features: 29

Columns:
['age', 'gender', 'marital_status', 'education', 'monthly_salary', 'employment_type', 'years_of_employment', 'company_type', 'house_type', 'monthly_rent', 'family_size', 'dependents', 'school_fees', 'college_fees', 'travel_expenses', 'groceries_utilities', 'other_monthly_expenses', 'existing_loans', 'current_emi_amount', 'credit_score', 'bank_balance', 'emergency_fund', 'emi_scenario', 'requested_amount', 'requested_tenure', 'max_monthly_emi', 'total_monthly_expenses', 'disposable_income', 'expense_to_income_ratio']

Data types:
age                        float64
gender                       int64
marital_status               int64
education                    int64
monthly_salary             float64
employment_type              int64
years_of_employment        float64
company_type                 int64
house_type                   int64
monthly_rent               float64
family_size                  int64
dependents                   int64
school_fees 

In [30]:
for col in [
    "gender",
    "marital_status",
    "education",
    "employment_type",
    "company_type",
    "house_type",
    "emi_scenario"
]:
    print("\n", col)
    print(df[col].value_counts().sort_index())


 gender
gender
0      1131
1      1108
2    151835
3      1771
4      1789
5    227755
6      1101
7      1743
Name: count, dtype: int64

 marital_status
marital_status
0    295393
1     92840
Name: count, dtype: int64

 education
education
0    174628
1     58571
2     96782
3     58252
Name: count, dtype: int64

 employment_type
employment_type
0     77819
1    271494
2     38920
Name: count, dtype: int64

 company_type
company_type
0    116095
1     97260
2     97192
3     19384
4     58302
Name: count, dtype: int64

 house_type
house_type
0     96746
1    136457
2    155030
Name: count, dtype: int64

 emi_scenario
emi_scenario
0    77618
1    77630
2    77632
3    77715
4    77638
Name: count, dtype: int64


In [ ]:
# Predict
y_pred_hgb = hgb_model.predict(X_test)

print("Predictions completed!")

In [ ]:
# Evaluate
hgb_accuracy = accuracy_score(y_test, y_pred_hgb)

print("HistGradientBoosting Accuracy:", hgb_accuracy)
print("Accuracy (%):", hgb_accuracy * 100)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_hgb))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_hgb))

**HistGradientBoosting model is clearly the best so far with the highest accuracy = 98.81%**

# **XGBoost Classifier**

In [ ]:
# Import XGBoost
from xgboost import XGBClassifier

In [ ]:
# Train XGBoost classifier
xgb_classifier = XGBClassifier(
    n_estimators=200,
    learning_rate=0.05,
    max_depth=6,
    random_state=42,
    n_jobs=-1,
    eval_metric="logloss"
)

xgb_classifier.fit(
    X_train,
    y_train
)

print("XGBoost Classifier training completed!")

In [ ]:
# Make Prediction
# XGBoost was trained using the original feature columns
X_test_encoded = X_test

y_pred_xgb = xgb_classifier.predict(X_test_encoded)

y_prob_xgb = xgb_classifier.predict_proba(X_test_encoded)[:, 1]

print("Predictions completed!")

In [ ]:
# Evaluate
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    classification_report
)

accuracy_xgb = accuracy_score(y_test, y_pred_xgb)

precision_xgb = precision_score(
    y_test,
    y_pred_xgb,
    average="weighted"
)

recall_xgb = recall_score(
    y_test,
    y_pred_xgb,
    average="weighted"
)

f1_xgb = f1_score(
    y_test,
    y_pred_xgb,
    average="weighted"
)

# Multiclass ROC-AUC
y_prob_xgb_multiclass = xgb_classifier.predict_proba(X_test_encoded)

roc_auc_xgb = roc_auc_score(
    y_test,
    y_prob_xgb_multiclass,
    multi_class="ovr",
    average="weighted"
)

print("XGBoost Classifier Results")
print("==========================")
print("Accuracy :", accuracy_xgb)
print("Precision:", precision_xgb)
print("Recall   :", recall_xgb)
print("F1 Score :", f1_xgb)
print("ROC-AUC  :", roc_auc_xgb)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_xgb))

# Create the final model comparison

In [ ]:
import pandas as pd
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

# Function to calculate multiclass metrics
def calculate_metrics(y_true, y_pred, model_name):
    
    accuracy = accuracy_score(y_true, y_pred)
    
    precision = precision_score(
        y_true,
        y_pred,
        average="weighted",
        zero_division=0
    )
    
    recall = recall_score(
        y_true,
        y_pred,
        average="weighted",
        zero_division=0
    )
    
    f1 = f1_score(
        y_true,
        y_pred,
        average="weighted",
        zero_division=0
    )
    
    return {
        "Model": model_name,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1_Score": f1
    }


# Calculate metrics for all available models
results = []

results.append(
    calculate_metrics(
        y_test,
        y_pred_logistic,
        "Logistic Regression"
    )
)

results.append(
    calculate_metrics(
        y_test,
        y_pred_rf,
        "Random Forest Classifier"
    )
)

results.append(
    calculate_metrics(
        y_test,
        y_pred_xgb,
        "XGBoost Classifier"
    )
)

results.append(
    calculate_metrics(
        y_test,
        y_pred_hgb,
        "HistGradientBoosting"
    )
)


classification_results = pd.DataFrame(results)

classification_results = classification_results.sort_values(
    by="F1_Score",
    ascending=False
).reset_index(drop=True)

display(classification_results)

In [ ]:
# ROC-AUC
y_prob_xgb

In [ ]:
# Select HistGradientBoosting as the final classifier
best_model = hgb_model

print("Best Classification Model: HistGradientBoostingClassifier")
print("Test Accuracy:", hgb_accuracy)
print("Test Accuracy (%):", hgb_accuracy * 100)

In [ ]:
# Target Encoder
from sklearn.preprocessing import LabelEncoder
import joblib

target_encoder = LabelEncoder()

# Original target labels from the cleaned dataframe
target_encoder.fit(df["emi_eligibility"])

print("Target classes:")
print(target_encoder.classes_)

In [ ]:
# Save Target Encoder
joblib.dump(
    target_encoder,
    "target_encoder.pkl"
)

print("Target encoder saved successfully!")

In [ ]:
# Save the model
import joblib

joblib.dump(
    best_model,
    "best_classification_model.pkl"
)

print("Best classification model saved successfully!")


In [ ]:
# Final verification
import os

print("Model file exists:",
      os.path.exists("best_classification_model.pkl"))

print("Target encoder exists:",
      os.path.exists("target_encoder.pkl"))